### ISSUE TITLE
Is there a way to query whether a given tag is supported by Exiv2?

### ISSUE THREAD

**Issue by tester0077** [Exiv2/exiv2#987]
In trying to port an existing and no longer actively maintained project, I have run into an issue where my fork, based on the existing code tries to read data for a tag "Exif.MinoltaCsNew.ISOSpeed".
Presumably, the code I am working with was initially meant for & compiled against an earlier version of Exiv2lib and now that it is linked with 0.2.72 it throws exceptions. Or perhaps the exceptions were simply ignored.
When Exiv2 0.2.72 throws the exception, the current code skips testing for any other tags for the given image file - not what I want or need.
Looking through the current Exiv2 code, I cannot find this string and hence I get the exception - no real problem in that   ... except I would expect that possibly at some time in past versions it was valid and no longer is. (some perusal of the release notes shows that potentially this change might date back to changes from 0.20 to 0.21)
Naturally, I will have to update these no-longer supported tags, but I would like to try and make the current code give a better response for such cases where built-in assumptions about libraries not under my direct control are modified and result in unexpected results and behaviour.

Hence my question: is there a way to check whether any tags I might be wanting to read are actually 'valid' or will I have to rely on simply verifying that the exiv2lib to be linked in is of a specific version and/or handle all exceptions for each tag I want to read?
Any recommendations or pointer much appreciated :-)

**Comment by clanmills** [comment]
Oh gosh, Arnold.  You always ask tough questions that require research.

It's unlikely that you will find the string "Exif.MinoltaCsNew.ISOSpeed" outside the test suite.  The tag means this is part of the Minolta makernote tag and relates to camera settings.

Alison and I are having a mini-vacation on the Isle of Wight.  I'll look at this and give you an update at the weekend.

**Comment by tester0077** [comment]
No rush -Just trying to educate myself.  I have simply disabled that particular part of the code - does not seem to be used now and possibly was part of initial testing. Don't get any answers from the original developer - so I have to second guess ....
FWIW, my e-mail service is down because the ISP - Telus - has major issues and had been having them since last Thursday - no email service for nearly a week for both BC & Alberta customers.
I can just imagine what business customers are experiencing :-(
But I do have another question on the 'stack' for when e-mail is up again

**Comment by tester0077** [comment]
Another FWIW: :)
The link on your  https://www.exiv2.org/makernote.html#R5 to http://www.dalibor.cz/minolta/makernote.htm is dead and was replaced by http://www.dalibor.cz/software

**Comment by clanmills** [comment]
There's a sample program `taglist` that generates all the tags supported by Exiv2.  The output from that is used to generate the web pages on exiv2.org relating to metadata and makernotes.

I believe the Exif and IPTC defined valid metadata names and they are supported.  XMP is "Extensible" and you can define your own tags (and arrays and objects).

Thanks for bring the broken link to my attention.  I'll fix that when I update the web-site for v0.27.3 which is scheduled for 2019-09-30.

Have a look at `taglist` and we'll talk more when I get home.

**Comment by clanmills** [comment]
Arnold:
```
$ taglist --help
Usage: taglist [--group name|Groups|Exif|Canon|CanonCs|CanonSi|CanonCf|Fujifilm|Minolta|Nikon1|Nikon2|Nikon3|Olympus|Panasonic|Pentax|Sigma|Sony|Iptc|dc|xmp|xmpRights|xmpMM|xmpBJ|xmpTPg|xmpDM|pdf|photoshop|crs|tiff|exif|aux|iptc]
Print Exif tags, MakerNote tags, or Iptc datasets
$ 
```
Let me explain what the tag names mean.

Tag: Family.Group.Tag
Family: Exif | Iptc | Xmp
Group : There are 106 groups:
```
$ taglist Groups | wc
     106     106    1016
$ taglist Groups | grep Minolta
Minolta
MinoltaCs5D
MinoltaCs7D
MinoltaCsOld
MinoltaCsNew
SonyMinolta
$
```
Tag: Can be almost anything and depend on the Group.
```
$ taglist MinoltaCsNew
ExposureMode,	1,	0x0001,	MinoltaCsNew,	Exif.MinoltaCsNew.ExposureMode,	Long,	Exposure mode
FlashMode,	2,	0x0002,	MinoltaCsNew,	Exif.MinoltaCsNew.FlashMode,	Long,	Flash mode
WhiteBalance,	3,	0x0003,	MinoltaCsNew,	Exif.MinoltaCsNew.WhiteBalance,	Long,	White balance
ImageSize,	4,	0x0004,	MinoltaCsNew,	Exif.MinoltaCsNew.ImageSize,	Long,	Image size
Quality,	5,	0x0005,	MinoltaCsNew,	Exif.MinoltaCsNew.Quality,	Long,	Image quality
DriveMode,	6,	0x0006,	MinoltaCsNew,	Exif.MinoltaCsNew.DriveMode,	Long,	Drive mode
MeteringMode,	7,	0x0007,	MinoltaCsNew,	Exif.MinoltaCsNew.MeteringMode,	Long,	Metering mode
ISO,	8,	0x0008,	MinoltaCsNew,	Exif.MinoltaCsNew.ISO,	Long,	ISO Value
ExposureTime,	9,	0x0009,	MinoltaCsNew,	Exif.MinoltaCsNew.ExposureTime,	Long,	Exposure time
FNumber,	10,	0x000a,	MinoltaCsNew,	Exif.MinoltaCsNew.FNumber,	Long,	The F-Number
MacroMode,	11,	0x000b,	MinoltaCsNew,	Exif.MinoltaCsNew.MacroMode,	Long,	Macro mode
DigitalZoom,	12,	0x000c,	MinoltaCsNew,	Exif.MinoltaCsNew.DigitalZoom,	Long,	Digital zoom
ExposureCompensation,	13,	0x000d,	MinoltaCsNew,	Exif.MinoltaCsNew.ExposureCompensation,	Long,	Exposure compensation
BracketStep,	14,	0x000e,	MinoltaCsNew,	Exif.MinoltaCsNew.BracketStep,	Long,	Bracket step
IntervalLength,	16,	0x0010,	MinoltaCsNew,	Exif.MinoltaCsNew.IntervalLength,	Long,	Interval length
IntervalNumber,	17,	0x0011,	MinoltaCsNew,	Exif.MinoltaCsNew.IntervalNumber,	Long,	Interval number
FocalLength,	18,	0x0012,	MinoltaCsNew,	Exif.MinoltaCsNew.FocalLength,	Long,	Focal length
FocusDistance,	19,	0x0013,	MinoltaCsNew,	Exif.MinoltaCsNew.FocusDistance,	Long,	Focus distance
FlashFired,	20,	0x0014,	MinoltaCsNew,	Exif.MinoltaCsNew.FlashFired,	Long,	Flash fired
MinoltaDate,	21,	0x0015,	MinoltaCsNew,	Exif.MinoltaCsNew.MinoltaDate,	Long,	Minolta date
MinoltaTime,	22,	0x0016,	MinoltaCsNew,	Exif.MinoltaCsNew.MinoltaTime,	Long,	Minolta time
MaxAperture,	23,	0x0017,	MinoltaCsNew,	Exif.MinoltaCsNew.MaxAperture,	Long,	Max aperture
FileNumberMemory,	26,	0x001a,	MinoltaCsNew,	Exif.MinoltaCsNew.FileNumberMemory,	Long,	File number memory
LastFileNumber,	27,	0x001b,	MinoltaCsNew,	Exif.MinoltaCsNew.LastFileNumber,	Long,	Last image number
ColorBalanceRed,	28,	0x001c,	MinoltaCsNew,	Exif.MinoltaCsNew.ColorBalanceRed,	Long,	Color balance red
ColorBalanceGreen,	29,	0x001d,	MinoltaCsNew,	Exif.MinoltaCsNew.ColorBalanceGreen,	Long,	Color balance green
ColorBalanceBlue,	30,	0x001e,	MinoltaCsNew,	Exif.MinoltaCsNew.ColorBalanceBlue,	Long,	Color balance blue
Saturation,	31,	0x001f,	MinoltaCsNew,	Exif.MinoltaCsNew.Saturation,	Long,	Saturation
Contrast,	32,	0x0020,	MinoltaCsNew,	Exif.MinoltaCsNew.Contrast,	Long,	Contrast
Sharpness,	33,	0x0021,	MinoltaCsNew,	Exif.MinoltaCsNew.Sharpness,	Long,	Sharpness
SubjectProgram,	34,	0x0022,	MinoltaCsNew,	Exif.MinoltaCsNew.SubjectProgram,	Long,	Subject program
FlashExposureComp,	35,	0x0023,	MinoltaCsNew,	Exif.MinoltaCsNew.FlashExposureComp,	Long,	Flash exposure compensation in EV
ISOSetting,	36,	0x0024,	MinoltaCsNew,	Exif.MinoltaCsNew.ISOSetting,	Long,	ISO setting
MinoltaModel,	37,	0x0025,	MinoltaCsNew,	Exif.MinoltaCsNew.MinoltaModel,	Long,	Minolta model
IntervalMode,	38,	0x0026,	MinoltaCsNew,	Exif.MinoltaCsNew.IntervalMode,	Long,	Interval mode
FolderName,	39,	0x0027,	MinoltaCsNew,	Exif.MinoltaCsNew.FolderName,	Long,	Folder name
ColorMode,	40,	0x0028,	MinoltaCsNew,	Exif.MinoltaCsNew.ColorMode,	Long,	ColorMode
ColorFilter,	41,	0x0029,	MinoltaCsNew,	Exif.MinoltaCsNew.ColorFilter,	Long,	Color filter
BWFilter,	42,	0x002a,	MinoltaCsNew,	Exif.MinoltaCsNew.BWFilter,	Long,	Black and white filter
Internal Flash,	43,	0x002b,	MinoltaCsNew,	Exif.MinoltaCsNew.Internal Flash,	Long,	Internal Flash
Brightness,	44,	0x002c,	MinoltaCsNew,	Exif.MinoltaCsNew.Brightness,	Long,	Brightness
SpotFocusPointX,	45,	0x002d,	MinoltaCsNew,	Exif.MinoltaCsNew.SpotFocusPointX,	Long,	Spot focus point X
SpotFocusPointY,	46,	0x002e,	MinoltaCsNew,	Exif.MinoltaCsNew.SpotFocusPointY,	Long,	Spot focus point Y
WideFocusZone,	47,	0x002f,	MinoltaCsNew,	Exif.MinoltaCsNew.WideFocusZone,	Long,	Wide focus zone
FocusMode,	48,	0x0030,	MinoltaCsNew,	Exif.MinoltaCsNew.FocusMode,	Long,	Focus mode
FocusArea,	49,	0x0031,	MinoltaCsNew,	Exif.MinoltaCsNew.FocusArea,	Long,	Focus area
DECPosition,	50,	0x0032,	MinoltaCsNew,	Exif.MinoltaCsNew.DECPosition,	Long,	DEC switch position
ColorProfile,	51,	0x0033,	MinoltaCsNew,	Exif.MinoltaCsNew.ColorProfile,	Long,	Color profile
DataImprint,	52,	0x0034,	MinoltaCsNew,	Exif.MinoltaCsNew.DataImprint,	Long,	Data Imprint
FlashMetering,	63,	0x003f,	MinoltaCsNew,	Exif.MinoltaCsNew.FlashMetering,	Long,	Flash metering
$ 
```
There isn't a tag Exif.MinoltaCsNew.ISOSpeed.  There is a Exif.MinoltaCSNew.ISO

You can use the program exifvalue to look for a tag in a file.  If the tag doesn't exist in the file, it will report "value not set":
```
$ exifvalue ~/Stonehenge.jpg Exif.MinoltaCsNew.ISO
Caught Exiv2 exception 'Value not set'
$
``` 
If the tag is not known, it will report 'Invalid tag':
```
$ exifvalue ~/Stonehenge.jpg Exif.MinoltaCsNew.ISOSpeed
Caught Exiv2 exception 'Invalid tag name or ifdId `ISOSpeed', ifdId 37'
$
```

Is there a way to report every tag known to exiv2?  Yes.  There are 5430 known tags:

```
$ for group in $(taglist Groups); do for tag in $(taglist $group | cut -d, -f 1) ; do echo $group.$tag ; done; done
Image.ProcessingSoftware
Image.NewSubfileType
Image.SubfileType
Image.ImageWidth
...
$ for group in $(taglist Groups); do for tag in $(taglist $group | cut -d, -f 1) ; do echo $group.$tag ; done; done | wc
    5430    5430  130555
$
```

Now, let me explain why there are 106 groups.  There are about 10 camera manufacturers (Canon, Minolta, Nikon etc) and they use the tag Exif.Photo.MakerNote to store data in a myriad of different (and proprietary standards).

```
$ exifvalue ~/Stonehenge.jpg Exif.Photo.MakerNote
78 105 107 111 110 0 2 ...
```

Exiv2 has code to read/modify/write makernotes.  All achieved by reverse engineering.  References on the web site.

The MakerNote usually isn't a simple structure.  The manufacturer usually has "sub-records" for Camera Settings (Cs), AutoFocus (Af) and so on.  Additionally, t ehe format of the sub-records can evolve and change with time.  For example (as above)
```
$ taglist Groups | grep Minolta
Minolta
MinoltaCs5D
MinoltaCs7D
MinoltaCsOld
MinoltaCsNew
SonyMinolta
$
```
So, Minolta have 6 "sub-records".  Other manufacturers have more.  Let's say 10 manufacturers have an average of 10 "sub-records".  That's 100 groups.

Now to address your concern about `Exif.MinoltaCsNew.ISOSpeed`.  It will throw an exception in Exiv2 v0.27.2.  Was it defined in an earlier version of Exiv2 such as 0.21?  I don't know.

Your application code has to use exception handlers to catch these matters and determine what to do.  Without getting involved with your application code, I can't comment on your best approach to dealing with this.  There is a macro EXIV2_TEST_VERSION which enables you to have version specific code in your application.

I am writing a book "Image Metadata and Exiv2 Architecture" and will include this information.  I intend to run a work-shop based on the book at LGM in Rennes in 2020. #911

I'm going to close this issue, as I believe I have answered your question.  I'm happy to continue the discussion if you wish.  However, please don't consume the time that I would prefer to spend on the book.

**Comment by clanmills** [comment]
Arnold:
```
$ taglist --help
Usage: taglist [--group name|Groups|Exif|Canon|CanonCs|CanonSi|CanonCf|Fujifilm|Minolta|Nikon1|Nikon2|Nikon3|Olympus|Panasonic|Pentax|Sigma|Sony|Iptc|dc|xmp|xmpRights|xmpMM|xmpBJ|xmpTPg|xmpDM|pdf|photoshop|crs|tiff|exif|aux|iptc]
Print Exif tags, MakerNote tags, or Iptc datasets
$ 
```
Let me explain what the tag names mean.

Tag: Family.Group.Tag
Family: Exif | Iptc | Xmp
Group : There are 106 groups:
```
$ taglist Groups | wc
     106     106    1016
$ taglist Groups | grep Minolta
Minolta
MinoltaCs5D
MinoltaCs7D
MinoltaCsOld
MinoltaCsNew
SonyMinolta
$
```
Tag: Can be almost anything and depend on the Group.
```
$ taglist MinoltaCsNew
ExposureMode,	1,	0x0001,	MinoltaCsNew,	Exif.MinoltaCsNew.ExposureMode,	Long,	Exposure mode
FlashMode,	2,	0x0002,	MinoltaCsNew,	Exif.MinoltaCsNew.FlashMode,	Long,	Flash mode
WhiteBalance,	3,	0x0003,	MinoltaCsNew,	Exif.MinoltaCsNew.WhiteBalance,	Long,	White balance
ImageSize,	4,	0x0004,	MinoltaCsNew,	Exif.MinoltaCsNew.ImageSize,	Long,	Image size
Quality,	5,	0x0005,	MinoltaCsNew,	Exif.MinoltaCsNew.Quality,	Long,	Image quality
DriveMode,	6,	0x0006,	MinoltaCsNew,	Exif.MinoltaCsNew.DriveMode,	Long,	Drive mode
MeteringMode,	7,	0x0007,	MinoltaCsNew,	Exif.MinoltaCsNew.MeteringMode,	Long,	Metering mode
ISO,	8,	0x0008,	MinoltaCsNew,	Exif.MinoltaCsNew.ISO,	Long,	ISO Value
ExposureTime,	9,	0x0009,	MinoltaCsNew,	Exif.MinoltaCsNew.ExposureTime,	Long,	Exposure time
FNumber,	10,	0x000a,	MinoltaCsNew,	Exif.MinoltaCsNew.FNumber,	Long,	The F-Number
MacroMode,	11,	0x000b,	MinoltaCsNew,	Exif.MinoltaCsNew.MacroMode,	Long,	Macro mode
DigitalZoom,	12,	0x000c,	MinoltaCsNew,	Exif.MinoltaCsNew.DigitalZoom,	Long,	Digital zoom
ExposureCompensation,	13,	0x000d,	MinoltaCsNew,	Exif.MinoltaCsNew.ExposureCompensation,	Long,	Exposure compensation
BracketStep,	14,	0x000e,	MinoltaCsNew,	Exif.MinoltaCsNew.BracketStep,	Long,	Bracket step
IntervalLength,	16,	0x0010,	MinoltaCsNew,	Exif.MinoltaCsNew.IntervalLength,	Long,	Interval length
IntervalNumber,	17,	0x0011,	MinoltaCsNew,	Exif.MinoltaCsNew.IntervalNumber,	Long,	Interval number
FocalLength,	18,	0x0012,	MinoltaCsNew,	Exif.MinoltaCsNew.FocalLength,	Long,	Focal length
FocusDistance,	19,	0x0013,	MinoltaCsNew,	Exif.MinoltaCsNew.FocusDistance,	Long,	Focus distance
FlashFired,	20,	0x0014,	MinoltaCsNew,	Exif.MinoltaCsNew.FlashFired,	Long,	Flash fired
MinoltaDate,	21,	0x0015,	MinoltaCsNew,	Exif.MinoltaCsNew.MinoltaDate,	Long,	Minolta date
MinoltaTime,	22,	0x0016,	MinoltaCsNew,	Exif.MinoltaCsNew.MinoltaTime,	Long,	Minolta time
MaxAperture,	23,	0x0017,	MinoltaCsNew,	Exif.MinoltaCsNew.MaxAperture,	Long,	Max aperture
FileNumberMemory,	26,	0x001a,	MinoltaCsNew,	Exif.MinoltaCsNew.FileNumberMemory,	Long,	File number memory
LastFileNumber,	27,	0x001b,	MinoltaCsNew,	Exif.MinoltaCsNew.LastFileNumber,	Long,	Last image number
ColorBalanceRed,	28,	0x001c,	MinoltaCsNew,	Exif.MinoltaCsNew.ColorBalanceRed,	Long,	Color balance red
ColorBalanceGreen,	29,	0x001d,	MinoltaCsNew,	Exif.MinoltaCsNew.ColorBalanceGreen,	Long,	Color balance green
ColorBalanceBlue,	30,	0x001e,	MinoltaCsNew,	Exif.MinoltaCsNew.ColorBalanceBlue,	Long,	Color balance blue
Saturation,	31,	0x001f,	MinoltaCsNew,	Exif.MinoltaCsNew.Saturation,	Long,	Saturation
Contrast,	32,	0x0020,	MinoltaCsNew,	Exif.MinoltaCsNew.Contrast,	Long,	Contrast
Sharpness,	33,	0x0021,	MinoltaCsNew,	Exif.MinoltaCsNew.Sharpness,	Long,	Sharpness
SubjectProgram,	34,	0x0022,	MinoltaCsNew,	Exif.MinoltaCsNew.SubjectProgram,	Long,	Subject program
FlashExposureComp,	35,	0x0023,	MinoltaCsNew,	Exif.MinoltaCsNew.FlashExposureComp,	Long,	Flash exposure compensation in EV
ISOSetting,	36,	0x0024,	MinoltaCsNew,	Exif.MinoltaCsNew.ISOSetting,	Long,	ISO setting
MinoltaModel,	37,	0x0025,	MinoltaCsNew,	Exif.MinoltaCsNew.MinoltaModel,	Long,	Minolta model
IntervalMode,	38,	0x0026,	MinoltaCsNew,	Exif.MinoltaCsNew.IntervalMode,	Long,	Interval mode
FolderName,	39,	0x0027,	MinoltaCsNew,	Exif.MinoltaCsNew.FolderName,	Long,	Folder name
ColorMode,	40,	0x0028,	MinoltaCsNew,	Exif.MinoltaCsNew.ColorMode,	Long,	ColorMode
ColorFilter,	41,	0x0029,	MinoltaCsNew,	Exif.MinoltaCsNew.ColorFilter,	Long,	Color filter
BWFilter,	42,	0x002a,	MinoltaCsNew,	Exif.MinoltaCsNew.BWFilter,	Long,	Black and white filter
Internal Flash,	43,	0x002b,	MinoltaCsNew,	Exif.MinoltaCsNew.Internal Flash,	Long,	Internal Flash
Brightness,	44,	0x002c,	MinoltaCsNew,	Exif.MinoltaCsNew.Brightness,	Long,	Brightness
SpotFocusPointX,	45,	0x002d,	MinoltaCsNew,	Exif.MinoltaCsNew.SpotFocusPointX,	Long,	Spot focus point X
SpotFocusPointY,	46,	0x002e,	MinoltaCsNew,	Exif.MinoltaCsNew.SpotFocusPointY,	Long,	Spot focus point Y
WideFocusZone,	47,	0x002f,	MinoltaCsNew,	Exif.MinoltaCsNew.WideFocusZone,	Long,	Wide focus zone
FocusMode,	48,	0x0030,	MinoltaCsNew,	Exif.MinoltaCsNew.FocusMode,	Long,	Focus mode
FocusArea,	49,	0x0031,	MinoltaCsNew,	Exif.MinoltaCsNew.FocusArea,	Long,	Focus area
DECPosition,	50,	0x0032,	MinoltaCsNew,	Exif.MinoltaCsNew.DECPosition,	Long,	DEC switch position
ColorProfile,	51,	0x0033,	MinoltaCsNew,	Exif.MinoltaCsNew.ColorProfile,	Long,	Color profile
DataImprint,	52,	0x0034,	MinoltaCsNew,	Exif.MinoltaCsNew.DataImprint,	Long,	Data Imprint
FlashMetering,	63,	0x003f,	MinoltaCsNew,	Exif.MinoltaCsNew.FlashMetering,	Long,	Flash metering
$ 
```
There isn't a tag Exif.MinoltaCsNew.ISOSpeed.  There is a Exif.MinoltaCSNew.ISO

You can use the program exifvalue to look for a tag in a file.  If the tag doesn't exist in the file, it will report "value not set":
```
$ exifvalue ~/Stonehenge.jpg Exif.MinoltaCsNew.ISO
Caught Exiv2 exception 'Value not set'
$
``` 
If the tag is not known, it will report 'Invalid tag':
```
$ exifvalue ~/Stonehenge.jpg Exif.MinoltaCsNew.ISOSpeed
Caught Exiv2 exception 'Invalid tag name or ifdId `ISOSpeed', ifdId 37'
$
```

Is there a way to report every tag known to exiv2?  Yes.  There are 5430 known tags:

```
$ for group in $(taglist Groups); do for tag in $(taglist $group | cut -d, -f 1) ; do echo $group.$tag ; done; done
Image.ProcessingSoftware
Image.NewSubfileType
Image.SubfileType
Image.ImageWidth
...
$ for group in $(taglist Groups); do for tag in $(taglist $group | cut -d, -f 1) ; do echo $group.$tag ; done; done | wc
    5430    5430  130555
$
```

Now, let me explain why there are 106 groups.  There are about 10 camera manufacturers (Canon, Minolta, Nikon etc) and they use the tag Exif.Photo.MakerNote to store data in a myriad of different (and proprietary standards).

```
$ exifvalue ~/Stonehenge.jpg Exif.Photo.MakerNote
78 105 107 111 110 0 2 ...
```

Exiv2 has code to read/modify/write makernotes.  All achieved by reverse engineering.  References on the web site.

The MakerNote usually isn't a simple structure.  The manufacturer usually has "sub-records" for Camera Settings (Cs), AutoFocus (Af) and so on.  Additionally, t ehe format of the sub-records can evolve and change with time.  For example (as above)
```
$ taglist Groups | grep Minolta
Minolta
MinoltaCs5D
MinoltaCs7D
MinoltaCsOld
MinoltaCsNew
SonyMinolta
$
```
So, Minolta have 6 "sub-records".  Other manufacturers have more.  Let's say 10 manufacturers have an average of 10 "sub-records".  That's 100 groups.

Now to address your concern about `Exif.MinoltaCsNew.ISOSpeed`.  It will throw an exception in Exiv2 v0.27.2.  Was it defined in an earlier version of Exiv2 such as 0.21?  I don't know.

Your application code has to use exception handlers to catch these matters and determine what to do.  Without getting involved with your application code, I can't comment on your best approach to dealing with this.  There is a macro EXIV2_TEST_VERSION which enables you to have version specific code in your application.

I am writing a book "Image Metadata and Exiv2 Architecture" and will include this information.  I intend to run a work-shop based on the book at LGM in Rennes in 2020. #911

I'm going to close this issue, as I believe I have answered your question.  I'm happy to continue the discussion if you wish.  However, please don't consume the time that I would prefer to spend on the book.

**Comment by clanmills** [comment]
I've modified samples/taglist.cpp by adding the options `all` and `ALL` to make it easier to search for tags.  I'm updated README-SAMPLES.md appropriately. #989

These changes will be released in Exiv2 v0.27.3 which is on-track to be released on 2019-09-30.
